kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1218 @@
+impl<'a, C: 'static + Chip> DynamicProcessLoading for DynamicProcessLoader<'a, C> {
+    fn set_client(&self, client: &'static dyn DynamicProcessLoadingClient) {
+        self.client.set(client);
+    }
+
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode> {
+        //TODO(?): Check if it is a newer version of an existing app.
+        // We can potentially flash the new app and load it before erasing the old one.
+        // What happens to the process though? Need to delete it from the process array and load it back in?
+
+        let flash_start = self.flash.get().as_ptr() as usize; //start of the flash region
+        let flash = self.flash.get();
+
+        if self.state.get() == State::Idle {
+            match self.find_next_available_address(flash, app_length) {
+                Ok(new_app_start_address) => {
+                    let offset = new_app_start_address - flash_start;
+                    let new_process_flash = self
+                        .flash
+                        .get()
+                        .get(offset..offset + app_length)
+                        .ok_or(ErrorCode::FAIL)?;
+
+                    self.new_process_flash.set(new_process_flash);
+                    self.new_app_start_addr.set(new_app_start_address);
+                    self.new_app_length.set(app_length);
+
+                    match self.padding_requirement.get() {
+                        // if we decided we need to write a padding app after the new app, we go ahead and do it
+                        Some(PaddingRequirement::PostPad)
+                        | Some(PaddingRequirement::PreAndPostPad) => {
+                            // calculating the distance between our app and either the next app
+                            let new_app_end_address = new_app_start_address + app_length;
+                            let post_pad_length =
+                                self.next_app_start_addr.get() - new_app_end_address;
+
+                            let padding_result =
+                                self.write_padding_app(post_pad_length, new_app_end_address);
+                            let _ = match padding_result {
+                                Ok(()) => Ok(()),
+                                Err(e) => {
+                                    self.reset_process_loading_metadata();
+                                    Err(e) // this means we were unable to write the padding app
+                                }
+                            };
+                        }
+                        // Otherwise we let the client know we are done with the setup, and we are ready to write the app to flash
+                        None | Some(PaddingRequirement::PrePad) => {
+                            self.state.set(State::AppWrite);
+                            self.client.map(|client| {
+                                client.setup_done();
+                            });
```

Comment:
We can't do this. We can't call a callback from a downcall (ie from DynamicProcessLoading.setup).

This needs to just return something that says "setup complete" (as opposed to "setup is still pending") or this file needs a deferred call to signal the callback.